Code fixes and improvements

michalbundyra opened the pull request.

michalbundyra committed:
Eliminated "goto" instruction

michalbundyra committed:
Removed redundant semicolon after if statement

michalbundyra committed:
Import classes

michalbundyra committed:
Changes "strstr" to "strpos" - save memory

michalbundyra committed:
Format change - consistency

michalbundyra committed:
Added private properties in tests

michalbundyra committed:
Added `@param` tags in tests with dataProvider

michalbundyra committed:
Changed deprecated `setExpectedException` to `expectException` and `expectExceptionMessage`

michalbundyra committed:
Changed `assertTrue(file_exists())` to `assertFileExists()`

michalbundyra committed:
Removed redundant variable

michalbundyra committed:
Bumped PHPUnit to ^5.7 in composer.json

weierophinney commented on src/GenerateProgrammaticPipelineFromConfig/Generator.php:
```diff
@@ -241,32 +241,20 @@ private function generateRoutes(array $config)
                     ? sprintf(self::TEMPLATE_ROUTED_NO_METHOD_NO_NAME, $path, $middleware)
                     : sprintf(self::TEMPLATE_ROUTED_NO_METHOD_WITH_NAME, $path, $middleware, $spec['name']);
-
-                goto options;
```
I know many folks do not like goto, but in this case, it eliminated a lot of code duplication. I find the revised version quite more cumbersome to read, due to additional nesting and duplication of the `$route` assigment ternary across three conditions.

michalbundyra replied:
I'm one of these folks who doesn't like `goto` :)

`if..elseif..else` statement doesn't produce more code there. There is no any duplication right now, as it was also with `goto` instruction. So basicaly now this 'if...else` statement is to get correct `route` - so I think it should be extracted to separate method, as I just did in 8088334.

Do you still think it is harder to understand and less maintainable?

weierophinney commented on test/GenerateProgrammaticPipelineFromConfig/GeneratorTest.php:
```diff
@@ -77,7 +83,9 @@ public function testGeneratesExpectedPipeline()
-        $this->setExpectedException(GeneratorException::class, 'not found');
+
+        $this->expectException(GeneratorException::class);
+        $this->expectExceptionMessage('not found');
```
This seems overly verbose, compared to the original. What prompted this change, exactly?

michalbundyra replied:
Since PHPUnit 5.2 `setExpectedException` method is deprecated. PHPUnit is development requirements so it's better to avoid using deprecated methods. It will be worth in future when we want migrate to newer version, where this method will be removed.

weierophinney replied:
I was unaware of that; thanks for pointing it out!

weierophinney commented on test/ScanForErrorMiddleware/CommandTest.php:
```diff
@@ -8,8 +8,10 @@
 namespace ZendTest\Expressive\Tooling\ScanForErrorMiddleware;
 
 use org\bovigo\vfs\vfsStream;
+use org\bovigo\vfs\vfsStreamDirectory;
```
Why are you importing this? It's not referenced anywhere...

michalbundyra replied:
Line 38: `/** @var vfsStreamDirectory */`

weierophinney commented on test/ScanForErrorMiddleware/ErrorMiddlewareFilterTest.php:
```diff
@@ -10,6 +10,7 @@
 use ArrayIterator;
 use DirectoryIterator;
 use org\bovigo\vfs\vfsStream;
+use org\bovigo\vfs\vfsStreamDirectory;
```
Same question here: it's not used anywhere, and tests pass without importing it...

michalbundyra replied:
Line 22: `/** @var vfsStreamDirectory */`. We can use there FQCN but I think it's better when it's imported. Then in IDE we can get suggestions etc.

weierophinney replied:
IDEs can get suggestions when you use a FQCN as well:

```php
/** @var \org\bovigo\vfs\vfsStreamDirectory */
```

I'd argue we could do the same with `ProphecyInterface`. Since they're only being used in annotations for typehints, I'd prefer that.

weierophinney commented on test/ScanForErrorMiddleware/ScannerTest.php:
```diff
@@ -10,8 +10,10 @@
 use Countable;
 use IteratorAggregate;
 use org\bovigo\vfs\vfsStream;
+use org\bovigo\vfs\vfsStreamDirectory;
```
ditto

michalbundyra replied:
Line 87: `/** @var vfsStreamDirectory */`

weierophinney commented on test/ScanForErrorMiddleware/ScannerTest.php:
```diff
@@ -125,10 +139,13 @@ public function testScanningEmitsInfoToConsoleWhenEncounteringFilesOfInterest()
             ->writeLine(
                 Argument::that(function ($arg) {
-                    if (! strstr($arg, 'src/ErrorMiddleware.php')) {
+                    if (false === strpos($arg, 'src/ErrorMiddleware.php')) {
```
The original was correct; `strstr` returns boolean `false` if the needle is not found in the haystack. The only reason it might emit a false positive is if it matches exactly at the beginning, as it would return `''` in that case; in this particular comparison, that would never happen...

michalbundyra replied:
As far as I know `strpos` is more efficient, uses less memory because it returns `int` or `false`. `strstr` returns portion of string  or `false`. It's just a good practice to use `strpos` instead (when possible).

weierophinney commented on test/ScanForErrorMiddleware/ScannerTest.php:
```diff
@@ -125,10 +139,13 @@ public function testScanningEmitsInfoToConsoleWhenEncounteringFilesOfInterest()
                         return false;
                     }
-                    if (! strstr($arg, sprintf('<error>implementing %s</error>', ErrorMiddlewareInterface::class))) {
+                    if (false === strpos(
```
On the flip side, this one _would_ occur, so the proposed change is more correct.

weierophinney replied:
On consideration, changing all occurrences of `! strstr` to `false === strpos` would make it more internally consistent, so I'm fine with these changes.

weierophinney reviewed (changes requested):
Overall, this looks fine. However, I have concerns about two sets of changes:

- The change to eliminate `goto` usage leads to nested conditionals and duplication of a unwieldy ternary... which was the original impetus for using `goto` in this particular case. I'm not convinced that the changes make the code easier to understand or more maintainable.
- Replacing a single `setExpectedException($type, $messageContains)` call with a set of `expectException()` + `expectExceptionMessage()` calls  also feels more verbose, and I'm not convinced that doing so makes the code clearer.

Other than those, the other proposed changes make sense to me.

michalbundyra committed:
Extract method "getRoute" from spec array

- previously it was part with "goto" instruction,
extreacted to separate method to improve readability and maintainable
and also to eliminate "goto" instruction

weierophinney commented on src/GenerateProgrammaticPipelineFromConfig/Generator.php:
```diff
@@ -272,6 +248,40 @@ private function generateRoutes(array $config)
+     * @param array $spec
+     * @return string
+     */
+    private function getRoute(array $spec)
```
much better! Thanks!

weierophinney commented:
Thanks, @webimpress
